include/pybind11/pytypes.h

```diff
@@ -235,8 +247,10 @@ class accessor : public object_api<accessor<Policy>> {
     void operator=(const accessor &a) & { operator=(handle(a)); }
     void operator=(const object &o) && { std::move(*this).operator=(handle(o)); }
     void operator=(const object &o) & { operator=(handle(o)); }
-    void operator=(handle value) && { Policy::set(obj, key, value); }
-    void operator=(handle value) & { get_cache() = reinterpret_borrow<object>(value); }
+    template <typename T, enable_if_t<!std::is_base_of<object, T>::value && !std::is_base_of<accessor, T>::value, int> = 0>
+    void operator=(const T &value) && { Policy::set(obj, key, object_or_cast(value)); }
+    template <typename T, enable_if_t<!std::is_base_of<object, T>::value && !std::is_base_of<accessor, T>::value, int> = 0>
+    void operator=(const T value) & { get_cache() = reinterpret_borrow<object>(object_or_cast(value)); }
```

Comment:
I believe the `const object &` overload can be removed because the template will take care of it. The `const accessor &` overload must stay in order to override the default assignment operator (templates are not allowed to replace default compiler generated assignments).

The SFINAE can go away because it's handled by normal overload resolution (given an exact match between a template and a non-template function, the non-template function always has priority).

Forward `value` in the template.